File: erpmax/erpmax/page/report_builder/report_builder.py
# Copyright (c) 2024, ERPMax and Contributors
# License: MIT. See LICENSE

import frappe
from frappe import _


def get_context(context):
    """Get context for report builder page"""
    context.no_cache = 1
    context.title = _("Report Builder")
    
    # Get available doctypes
    context.doctypes = get_available_doctypes()
    
    # Get saved reports
    context.saved_reports = get_saved_reports()
    
    return context


def get_available_doctypes():
    """Get list of doctypes available for reporting"""
    doctypes = frappe.get_all(
        "DocType",
        filters={
            "istable": 0,
            "issingle": 0,
            "module": ["in", ["ERPMax", "Accounting", "Sales", "Purchase", 
                             "Inventory", "Commerce", "Banking", "Expense Management"]]
        },
        fields=["name", "module"],
        order_by="name asc"
    )
    
    return doctypes


def get_saved_reports():
    """Get list of saved custom reports"""
    return frappe.get_all(
        "Custom Report",
        fields=["name", "report_name", "doc_type", "owner", "modified"],
        order_by="modified desc",
        limit=20
    )


@frappe.whitelist()
def get_doctype_fields(doctype):
    """Get fields for a doctype"""
    meta = frappe.get_meta(doctype)
    
    fields = []
    for field in meta.fields:
        fields.append({
            "fieldname": field.fieldname,
            "fieldtype": field.fieldtype,
            "label": field.label,
            "options": field.options
        })
    
    return fields


@frappe.whitelist()
def execute_custom_report(config):
    """Execute a custom report"""
    import json
    
    if isinstance(config, str):
        config = json.loads(config)
    
    doctype = config.get("doctype")
    filters = config.get("filters", {})
    columns = config.get("columns", [])
    sorts = config.get("sorts", [])
    groups = config.get("groups", [])
    
    # Build query
    fields = [col.get("fieldname") for col in columns]
    
    # Execute query
    data = frappe.get_all(
        doctype,
        filters=filters,
        fields=fields,
        order_by=sorts,
        limit=1000
    )
    
    return {
        "columns": columns,
        "data": data,
        "total": len(data)
    }


@frappe.whitelist()
def save_report(config, report_name):
    """Save a custom report"""
    import json
    
    if isinstance(config, str):
        config = json.loads(config)
    
    # Create or update Custom Report
    report = frappe.get_doc({
        "doctype": "Custom Report",
        "report_name": report_name,
        "doc_type": config.get("doctype"),
        "config": json.dumps(config),
        "owner": frappe.session.user
    })
    
    report.insert()
    frappe.db.commit()
    
    return report.name


@frappe.whitelist()
def delete_report(report_name):
    """Delete a custom report"""
    frappe.delete_doc("Custom Report", report_name)
    frappe.db.commit()
    return True


@frappe.whitelist()
def generate_report(report_type):
    """Generate report by type"""
    try:
        if report_type == "trial_balance":
            from erpmax.accounting.report.trial_balance.trial_balance import execute
            columns, data = execute()
            return {"report_type": "Trial Balance", "data": data, "columns": columns}
        elif report_type == "general_ledger":
            from erpmax.accounting.report.general_ledger.general_ledger import execute
            columns, data = execute()
            return {"report_type": "General Ledger", "data": data, "columns": columns}
        elif report_type == "balance_sheet":
            from erpmax.accounting.report.balance_sheet.balance_sheet import execute
            columns, data = execute()
            return {"report_type": "Balance Sheet", "data": data, "columns": columns}
        elif report_type == "profit_and_loss":
            from erpmax.accounting.report.profit_and_loss.profit_and_loss import execute
            columns, data = execute()
            return {"report_type": "Profit and Loss", "data": data, "columns": columns}
        elif report_type == "sales_register":
            from erpmax.sales.report.sales_register.sales_register import execute
            columns, data = execute()
            return {"report_type": "Sales Register", "data": data, "columns": columns}
        elif report_type == "purchase_register":
            from erpmax.purchase.report.purchase_register.purchase_register import execute
            columns, data = execute()
            return {"report_type": "Purchase Register", "data": data, "columns": columns}
        elif report_type == "customer_summary":
            from erpmax.accounting.report.customer_summary.customer_summary import execute
            columns, data = execute()
            return {"report_type": "Customer Summary", "data": data, "columns": columns}
        elif report_type == "supplier_summary":
            from erpmax.accounting.report.supplier_summary.supplier_summary import execute
            columns, data = execute()
            return {"report_type": "Supplier Summary", "data": data, "columns": columns}
        else:
            return {"error": "Unknown report type: " + report_type}
    except Exception as e:
        return {"error": str(e)}

File: erpmax/erpmax/page/report_builder/report_builder.js
frappe.pages["report-builder"].on_page_load = function(wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __("Report Builder"),
        single_column: true,
    });

    const builder = new ReportBuilder(page);
    page.set_secondary_action(__("Refresh"), () => builder.reload(), "refresh");
};

class ReportBuilder {
    constructor(page) {
        this.page = page;
        this.render_skeleton();
        this.load_reports();
    }

    render_skeleton() {
        $(`
            <style>
                .rb-wrap { max-width: 1200px; margin: 0 auto; padding: 20px; }
                .rb-section { margin-bottom: 24px; }
                .rb-section-title { font-size: 14px; font-weight: 600; text-transform: uppercase; letter-spacing: 0.05em; color: var(--text-muted); margin-bottom: 12px; }
                .rb-report-grid { display: grid; grid-template-columns: repeat(auto-fill, minmax(200px, 1fr)); gap: 12px; }
                .rb-report-btn { background: var(--card-bg, #fff); border: 1px solid var(--border-color); border-radius: 8px; padding: 16px; cursor: pointer; text-align: center; transition: all 0.15s ease; }
                .rb-report-btn:hover { border-color: var(--primary); background: var(--primary-light, #f0f0ff); }
                .rb-report-btn.active { border-color: var(--primary); background: var(--primary-light, #f0f0ff); }
                .rb-report-icon { font-size: 24px; color: var(--primary); margin-bottom: 8px; }
                .rb-report-name { font-size: 14px; font-weight: 600; color: var(--text-color); }
                .rb-output { background: var(--card-bg, #fff); border: 1px solid var(--border-color); border-radius: 12px; padding: 20px; min-height: 300px; }
                .rb-loading { text-align: center; padding: 40px; color: var(--text-muted); }
                .rb-table { width: 100%; border-collapse: collapse; }
                .rb-table th, .rb-table td { border: 1px solid var(--border-color); padding: 10px 12px; text-align: left; font-size: 13px; }
                .rb-table th { background: var(--control-bg, #f5f5f5); font-weight: 600; text-transform: uppercase; letter-spacing: 0.03em; font-size: 11px; }
                .rb-table tr:hover { background: var(--control-bg, #f5f5f5); }
                .rb-table td { font-variant-numeric: tabular-nums; }
            </style>
            <div class="rb-wrap">
                <div class="rb-section">
                    <div class="rb-section-title">Available Reports</div>
                    <div class="rb-report-grid" id="rb-reports"></div>
                </div>
                <div class="rb-section">
                    <div class="rb-section-title">Report Output</div>
                    <div class="rb-output" id="rb-output">
                        <div class="rb-loading">Select a report to generate</div>
                    </div>
                </div>
            </div>
        `).appendTo(this.page.body);
    }

    load_reports() {
        const reports = [
            { name: "trial_balance", label: "Trial Balance", icon: "fa-balance-scale" },
            { name: "general_ledger", label: "General Ledger", icon: "fa-book" },
            { name: "balance_sheet", label: "Balance Sheet", icon: "fa-file-text" },
            { name: "profit_and_loss", label: "Profit & Loss", icon: "fa-line-chart" },
            { name: "sales_register", label: "Sales Register", icon: "fa-shopping-cart" },
            { name: "purchase_register", label: "Purchase Register", icon: "fa-truck" },
            { name: "customer_summary", label: "Customer Summary", icon: "fa-users" },
            { name: "supplier_summary", label: "Supplier Summary", icon: "fa-building" },
        ];

        let html = "";
        reports.forEach(r => {
            html += `
                <div class="rb-report-btn" data-report="${r.name}" onclick="current_builder.generate_report('${r.name}')">
                    <div class="rb-report-icon"><i class="fa ${r.icon}"></i></div>
                    <div class="rb-report-name">${r.label}</div>
                </div>
            `;
        });
        $("#rb-reports").html(html);
        window.current_builder = this;
    }

    async generate_report(report_name) {
        // Highlight selected
        $(".rb-report-btn").removeClass("active");
        $(`.rb-report-btn[data-report="${report_name}"]`).addClass("active");

        // Show loading
        $("#rb-output").html('<div class="rb-loading"><i class="fa fa-spinner fa-spin"></i> Generating report...</div>');

        try {
            const r = await frappe.xcall("erpmax.erpmax.page.report_builder.report_builder.generate_report", {
                report_type: report_name
            });
            
            if (r && r.error) {
                $("#rb-output").html(`<div class="rb-loading" style="color: #dc2626">Error: ${r.error}</div>`);
            } else if (r && r.data && r.data.length > 0) {
                this.render_table(r);
            } else {
                $("#rb-output").html('<div class="rb-loading">No data found for this report</div>');
            }
        } catch (e) {
            console.error("Report error:", e);
            $("#rb-output").html(`<div class="rb-loading" style="color: #dc2626">Error: ${e.message || "Failed to generate report"}</div>`);
        }
    }

    render_table(report) {
        let html = `<div style="margin-bottom:12px;font-weight:600">${report.report_type || "Report"}</div>`;
        html += '<table class="rb-table"><thead><tr>';
        
        if (report.columns) {
            report.columns.forEach(col => {
                html += `<th>${col.label || col.fieldname || col}</th>`;
            });
        } else if (report.data && report.data.length > 0) {
            Object.keys(report.data[0]).forEach(key => {
                html += `<th>${key}</th>`;
            });
        }
        
        html += '</tr></thead><tbody>';
        
        report.data.forEach(row => {
            html += '<tr>';
            if (report.columns) {
                report.columns.forEach(col => {
                    const field = col.fieldname || col;
                    html += `<td>${row[field] !== undefined ? row[field] : ""}</td>`;
                });
            } else {
                Object.values(row).forEach(val => {
                    html += `<td>${val !== undefined ? val : ""}</td>`;
                });
            }
            html += '</tr>';
        });
        
        html += '</tbody></table>';
        html += `<div style="margin-top:12px;font-size:12px;color:var(--text-muted)">Total: ${report.data.length} records</div>`;
        
        $("#rb-output").html(html);
    }

    reload() {
        this.load_reports();
        $("#rb-output").html('<div class="rb-loading">Select a report to generate</div>');
    }
}
